### PR TITLE
[MM-55153] Consolidate Desktop App API, use new contextBridge endpoints when available

### PR DESCRIPTION
**Pull request by devinbinnie**
#### Summary
This PR contains changes to consolidate our Desktop App specific code (wrt. messaging between the apps) into a single module, containing all the hooks that the Web App needs to be aware of to work with the Desktop App. It also contains the original compatibility code to work with older Desktop App versions while those are still supported. The use of the Messaging API can be removed once those Desktop App versions are no longer supported.

A few other small changes have been made to the webapp to better accommodate the new API. Those changes have been commented below.

The architecture of these changes is up for debate, wrt. how they are accessed in the Web App, but the overall use of the hooks on the `window.desktopAPI` object will be the way going forward if we want to interact with the Desktop App. This is a much clearer and more intuitive interface that should improve the development experience with the Desktop App and allow the Web App to be in charge of what it is compatible with, also allowing the Desktop App to advertise easily what it is capable of.

Note that we'll want to test these changes with the current Desktop App (and other supported versions) as well as the one providing the new API (see below)

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55153

#### Related PR
mattermost/desktop/pull/2900: Desktop App PR containing the new API. Will be needed to test between the two apps.

```release-note
NONE
```

### CODE REVIEW

File: webapp/channels/src/utils/desktopAPI.ts
Comment (devinbinnie) [inline webapp/channels/src/utils/desktopAPI.ts:47]: To avoid having the rest of the app care about whether we use `window.postMessage` or the new API, we wrap a class here containing a set of listeners we register to be called when the `message` event comes down.

All of this will go away once we remove `window.postMessage` permanently.

File: webapp/channels/src/utils/desktopAPI.ts
Comment (devinbinnie) [inline webapp/channels/src/utils/desktopAPI.ts:196]: This is where I contradict myself a bit to say that we don't want the rest of the app to be aware of `window.postMessage` :P

Unfortunately we still need to have the app have some control over when to deactivate these listeners, so we do it React-style here by just returning the "off" hook.

File: webapp/channels/src/utils/desktopAPI.ts
Comment (devinbinnie) [inline webapp/channels/src/utils/desktopAPI.ts:9]: This needs to change before we merge, as we should be getting these types from an npm package.

Comment (devinbinnie) [reply webapp/channels/src/utils/desktopAPI.ts:9]: @hmhealey Might need your help with this one.

File: webapp/channels/src/components/logged_in/logged_in.tsx
Comment (larkox) [inline webapp/channels/src/components/logged_in/logged_in.tsx:185]: This doesn't seem handled in the changes of this file. Has it been moved somewhere else?

File: webapp/channels/src/components/login/login.tsx
Comment (larkox) [inline webapp/channels/src/components/login/login.tsx:443]: Any reason not to move this to the desktop api, so everything that is to be removed is in the same place?

File: webapp/channels/src/utils/desktopAPI.ts
Comment (larkox) [inline webapp/channels/src/utils/desktopAPI.ts:175]: Again, I prefer the await nomenclature than the `then` one, but 0/5.

File: webapp/channels/src/utils/desktopAPI.ts
Comment (larkox) [inline webapp/channels/src/utils/desktopAPI.ts:42]: desktopApp and desktopAPI are two names too similar XD Not sure if it is worth a name change, but I got really confused about how this will work because I thought this was `window.desktopApp`, defined in `plugins/exports.js`.

**Review by larkox — APPROVED** [review]
LGTM. A few comments here and there, but nothing blocking.

File: webapp/channels/src/utils/desktopAPI.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktopAPI.ts:1]: Hehe, I am still way too used to camelCase :P
But I can change it for sure.

File: webapp/channels/src/components/desktop_auth_token.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/desktop_auth_token.tsx:68]: This is all from before the API went in. I think the token creation belongs more to the Web App than it does the Desktop App, so I'll leave it here for now.

File: webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:63]: I agree. Will update.

File: webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:74]: Yeah I was kind of wondering this myself. I wanted a way to turn off the listeners (as I mentioned in an above comment) but I didn't want to spend too much time designing it since this functionality won't be necessary with the legacy stuff.

I'm open to ideas on how to improve it.

File: webapp/channels/src/components/logged_in/index.ts
Comment (devinbinnie) [reply webapp/channels/src/components/logged_in/index.ts:47]: This function is literally only used for this Desktop App feature as far as I can tell. The Desktop App doesn't need the whole channel object and I think it's simpler to avoid passing it all up, so I changed it here.

File: webapp/channels/src/components/logged_in/logged_in.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/logged_in/logged_in.tsx:185]: Right here ;)
https://github.com/mattermost/mattermost/blob/d6fc7253397f24992b3109e40581d48074c227b2/webapp/channels/src/utils/desktopAPI.ts#L36

File: webapp/channels/src/components/login/login.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/login/login.tsx:443]: Nothing specifically, I guess I worry a bit about someone using the `setSessionExpired` call and getting this param attached unexpectedly. That's my only concern. 1/5 on keeping it here.

File: webapp/channels/src/utils/desktopAPI.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktopAPI.ts:175]: Thanks for catching these :)

File: webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx
Comment (larkox) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:74]: It is still a listener, so... why not listen?

File: webapp/channels/src/utils/desktopAPI.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktopAPI.ts:42]: Hehe I know. To be honest, I'd actually prefer to use `window.desktop`, but because of the legacy stuff on there it was confusing...

Open to suggestions on how to improve it :)

File: webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:74]: Well we're not returning the listener though, we're returning the function to turn it off, hence the name `off` :P

Comment (larkox) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:75]: The way most of this kind of things is done in mobile is the following:
- `DeviceEventEmitter.addListener` returns a `EmitterSubscription`, which has a `remove` function (among other things)

This at the end is the same... you are adding a subscription, and you are receiving the means to remove it.

So I would read it more or less fine if it was something like...
`const removeListener = addListenerOnBrowserHistoryStatusUpdated(updateButtons);`

But... yeah, not sure either. 0/5 on this for sure.

File: webapp/channels/src/utils/desktopAPI.ts
Comment (larkox) [reply webapp/channels/src/utils/desktopAPI.ts]: No good ideas on how to improve it :(

File: webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:75]: Gotcha. So `EventEmitter` supports `on` and `off`, which is where I got this from: https://nodejs.org/api/events.html#emitteroneventname-listener

Also 0/5.

Comment (larkox) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:75]: Then let's wait for @hmhealey review so he can break the tie. Or bring a better idea :stuck_out_tongue:

File: webapp/channels/src/components/logged_in/index.ts
Comment (hmhealey) [reply webapp/channels/src/components/logged_in/index.ts:47]: Not for this PR, but it would been nice if we grouped or marked these functions that are only used by the desktop API. I guess that's what that comment above the function was meant to do 😛

Comment (devinbinnie) [reply webapp/channels/src/components/logged_in/index.ts:47]: Going forward, we might want to make a bigger Desktop App module to have:
1) The main module containing the name/version, as well as any helpers to talk to the API/do backwards compatibility
2) The individual wrapped API calls
3) Any other Desktop-specific code (which is where I'd imagine that function would go)

We COULD setup the framework for that here, and then start moving stuff over later. 1/5 on that.

File: webapp/channels/src/utils/desktop_api.ts
Comment (hmhealey) [inline webapp/channels/src/utils/desktop_api.ts:156]: 0/5 I think I'd rather this be a named export rather than a default export since that feels inconsistent for some reason, but I doubt it makes any real difference

File: webapp/channels/src/utils/desktop_api.ts
Comment (hmhealey) [reply webapp/channels/src/utils/desktop_api.ts:166]: In a similar vein, I wonder if making `desktopAPI` a field of the `DesktopApp` class would be a good idea so that we could inject a mock easier without having to change things on the window. That's probably a moot point though as long as `DesktopApp` is a singleton

File: webapp/channels/src/plugins/export.js
Comment (hmhealey) [inline webapp/channels/src/plugins/export.js:109]: How are we expecting plugins to interact with the desktop API? I would've expected the web app to expose some compatibility-friendly version of the desktop API to them instead of them just getting it from `window.desktopAPI`.

If we don't make the compatibility-friendly versions of those methods part of the `DesktopApp` like I was thinking about elsewhere, do we actually need to expose the `DesktopApp` itself to plugins? At the moment, it looks like it mostly contains setup and glue code that they wouldn't need direct access to

File: webapp/channels/src/utils/desktop_api.ts
Comment (hmhealey) [inline webapp/channels/src/utils/desktop_api.ts:190]: Does `desktopAPI.onUserActivityUpdate` not return a cleanup function of its own?

I know we discussed making the desktop app responsible for final cleanup for the app when it's closed or refresh, but these are going to leak memory if the component that registers these callbacks is ever unmounted. Even components like `LoggedIn` and `ChannelController` are remounted when moving to areas like the System Console, so I think we need cleanup functions for the newer versions of these

File: webapp/channels/src/utils/desktop_api.ts
Comment (hmhealey) [inline webapp/channels/src/utils/desktop_api.ts:171]: Not for this PR, but I wonder what our plan going forward should be for eventually deprecating this old desktop compatibility code. While we're not checking specific version numbers here, it would be nice if we could eventually remove some of this in a few years

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:75]: Yeah that could work. I'll see what I can do.

File: webapp/channels/src/actions/notification_actions.jsx
Comment (devinbinnie) [reply webapp/channels/src/actions/notification_actions.jsx:316]: I'd be fine making sure the parameters are consistently ordered, but again other than protecting people from fat-fingering I don't see any value. I'd much rather just be prescriptive about the API and what we add, and I'd like to avoid having an object that could be misused.

Again, 3/5 on that.

File: webapp/channels/src/components/logged_in/logged_in.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/logged_in/logged_in.tsx:121]: I think this was here before. If you don't think this is necessary, we can remove it. 0/5

File: webapp/channels/src/components/login/login.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/login/login.tsx:243]: Definitely seems like a nit, but I see what you mean :P
Unfortunately the way this all works seems to be a bit janky anyways. I can rework for this one.

File: webapp/channels/src/utils/browser_history.tsx
Comment (devinbinnie) [reply webapp/channels/src/utils/browser_history.tsx:18]: `path` actually comes from the call to `browserHistory.push`. I'm actually trying to remember why that code is there, but I know there's a reason.

Either way, I think that code will get removed so we can just leave it as is for now.

File: webapp/channels/src/utils/browser_history.tsx
Comment (devinbinnie) [reply webapp/channels/src/utils/browser_history.tsx:26]: Actually I guess it's not necessary here because it's just a listener. I guess I wanted to save having a listener registered when it would be needed. Unfortunately running it means it'll default to the message passing case, so we'd be registering it unnecessarily.

1/5 on keeping it, but it's probably not a huge deal to just remove the check as well.

File: webapp/channels/src/utils/desktop_api.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:16]: Heh, I kinda wanted your help on this actually. Not sure the best way to break up this module, since most of the API calls don't rely on that class (except in the legacy case :P) so I just left it the way I originally named it.

Open to suggestions :)

File: webapp/channels/src/utils/desktop_api.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:166]: Also went back and forth on this. I notice we don't really use classes much in the web app so I didn't want to force most of the API methods to be a member (singleton or not)

But I felt I needed one to store a few fields (without just exporting variables to be read), so I'm kinda stuck in the middle as to how to proceed.

My gut reaction here is to actually split up the `DesktopApp` class and the API methods into 2 seperate files, thoughts?

File: webapp/channels/src/plugins/export.js
Comment (devinbinnie) [reply webapp/channels/src/plugins/export.js:109]: So the `DesktopApp` class contains the functionality to register legacy listeners, and it also has the name/version. I wanted to make sure plugins had access to that at a minimum.

I was interested in sharing the other methods as well, but as it stands I don't think any plugins use it currently. But I would like to keep the API consistent and belonging to the Web App so we can share it with the plugins.

File: webapp/channels/src/utils/desktop_api.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:190]: Yeah I mentioned this in the other PR, we need to return the `unregister` call, but unfortunately that will break all listeners on the same channel, which isn't ideal.

I did want to avoid making `on`/`off` for everything, but that might be the only way.

File: webapp/channels/src/utils/desktop_api.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:171]: In less than a few years I imagine, this should all get removed once the Desktop App's that use it aren't supported anymore.

File: webapp/channels/src/actions/notification_actions.jsx
Comment (hmhealey) [reply webapp/channels/src/actions/notification_actions.jsx:316]: I'd say preventing mistakes is pretty valuable, but I'm fine to leave this

File: webapp/channels/src/components/logged_in/logged_in.tsx
Comment (hmhealey) [reply webapp/channels/src/components/logged_in/logged_in.tsx:121]: I'd prefer to remove it if it doesn't really do anything. If nothing else, it'd be better to blank out the field by setting it to `undefined` rather than using `delete` since I think think those have slightly different semantics that I don't entirely understand

File: webapp/channels/src/utils/browser_history.tsx
Comment (hmhealey) [reply webapp/channels/src/utils/browser_history.tsx:26]: I'm good with leaving that. I was just curious if you had a specific goal in mind with regards to having or not having those checks everywhere

File: webapp/channels/src/utils/desktop_api.ts
Comment (hmhealey) [reply webapp/channels/src/utils/desktop_api.ts:166]: I'd actually prefer the API methods to be instance methods of a class. I think the desktop API is similar enough to something like `Client4` or `WebSocketClient` that keeping them in a class makes sense to me conceptually. Plus, if we end up wanting to add internal state to the API code or doing something fancy, starting with a class encapsulates that and lets us make any changes without affecting callers.

As for if that should be in `DestkopApp` or in a separate class, I'm 0/5 on that. Keeping them together does mean exposing more areas of the code to the initialization and any lower level logic (like querying the desktop version directly), but it might make sense anyway. I feel similarly about if they should be in two different files or not, but that's only possible if they're not in a single class.

File: webapp/channels/src/plugins/export.js
Comment (hmhealey) [reply webapp/channels/src/plugins/export.js:109]: I agree that the web app should access the methods in the same way as a plugin would, but I think that should be via `DesktopApp` or a different class as mentioned above

File: webapp/channels/src/utils/desktop_api.ts
Comment (hmhealey) [reply webapp/channels/src/utils/desktop_api.ts:190]: I don't suppose there's a way for the desktop API to also return an `off` when you call an `on`? If this could just `return window.desktopAPI.onUserActivityUpdate(listener)`, that seems ideal

File: webapp/channels/src/utils/desktop_api.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:166]: If we're feeling that it should be a class member then I'm good with that, definitely feels cleaner to me. That way we'll just work with this one object. I'll push everything into the class.

File: webapp/channels/src/utils/desktop_api.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:156]: Since we're doing the above to push it all into a class, meaning that we won't be exporting anything else, I think a default export is fine.
Will leave as-is.

File: webapp/channels/src/plugins/export.js
Comment (devinbinnie) [reply webapp/channels/src/plugins/export.js:109]: Will do the class design as suggested above.

File: webapp/channels/src/utils/desktop_api.ts
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:190]: That would require changes to the Desktop App side of things, and is another thing to maintain for every single listener call. Let me see if I can find an intelligent way to wrap it and if I can, we'll do that since that would obviously be better for the Web App side of things.

If not, I'll come back here and we'll figure out something else.

File: webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/global_header/left_controls/history_buttons/history_buttons.tsx:75]: So thinking on this I'm not sure just adding `useDesktopListener = useEffect` is really necessary. I suppose it makes things a little clearer, but seems like we're just adding some bloat for that.

I've added a comment for this case, but maybe this is just my bias but I think the code is relatively self-explanatory. `on<doAThing>` always returns `off<doAThing>` which stops the listener. That feels like a pretty common convention to me.

I think this is something we can iterate on going forward, but since a) this is only one case so far and b) we don't really have full context as to the use of listeners yet, I'd rather not prescribe anything.

I'm going to say 2/5 on this right now.

**Review by hmhealey — APPROVED** [review]

**Review by yasserfaraazkhan — APPROVED** [review]
tested this pr along with desktop changes https://github.com/mattermost/desktop/pull/2900#pullrequestreview-1762579843

**Pull request merged**
